### PR TITLE
[core][tile mode] Reduce cut-off labels (part 2)

### PR DESCRIPTION
**Pull request by pozdnyakov**
Now, the intersecting symbols are placed across all layers _symbol by symbol_ according to the following rules:

  1. First we look, which of the tile border(s) the symbol intersects and prioritize the the symbol placement accordingly (high priority -> low priority): `vertical & horizontal -> vertical -> horizontal`
  2. For the symbols that intersect the same tile border(s), assuming the tile border split symbol into several sections, we look at the minimal section length. The symbol with a larger minimal section length is placed first.
  3. For the symbols that intersect the same tile border(s), and have equal minimal section length, we look at the anchor coordinates.
  4. Finally, if all the previous criteria are the same, we look at the symbol key hashes.

Fixes https://github.com/mapbox/mapbox-gl-native-team/issues/223

### CODE REVIEW

File: render-test/runner.cpp [hunk 1]
```diff
@@ -775,9 +788,49 @@ void TestRunner::run(TestMetadata& metadata) {
                 metadata.errorMessage = "Failed rendering tile: " + util::toString(tileId);
                 return;
             }
-
             auto xOffset = getImageTileOffset(xDims, tileId.canonical.x, metadata.pixelRatio);
             auto yOffset = getImageTileOffset(yDims, tileId.canonical.y, metadata.pixelRatio);
+
+            const auto& placedSymbols = ctx.getFrontend().getRenderer()->getPlacedSymbolsData();
+            for (const auto& placedSymbol : placedSymbols) {
```
Comment (tmpsantos) [inline render-test/runner.cpp:795]: Move the logic of detecting the label collision to its own function?

File: src/mbgl/text/placement.cpp [hunk 2]
```diff
@@ -59,6 +59,93 @@ const CollisionGroups::CollisionGroup& CollisionGroups::get(const std::string& s
...
+    // See https://github.com/mapbox/mapbox-gl-native/issues/12683
+    bool alwaysShowText =
+        textAllowOverlap && (iconAllowOverlap || !(getBucket().hasIconData() || getBucket().hasSdfIconData()) ||
+                             getLayout().get<style::IconOptional>());
+    bool alwaysShowIcon =
+        iconAllowOverlap && (textAllowOverlap || !getBucket().hasTextData() || getLayout().get<style::TextOptional>());
+
+    bool hasIconTextFit = getLayout().get<IconTextFit>() != IconTextFitType::None;
+
+    optional<CollisionBoundaries> avoidEdges;
```
Comment (tmpsantos) [inline src/mbgl/text/placement.cpp:146]: These properties with public access should be all `const`, no?

File: src/mbgl/text/placement.cpp [hunk 3]
```diff
@@ -143,419 +230,388 @@ Point<float> calculateVariableLayoutOffset(style::SymbolAnchorType anchor,
...
+            }
+        };
 
-    std::vector<ProjectedCollisionBox> textBoxes;
-    std::vector<ProjectedCollisionBox> iconBoxes;
+        // Line or point label placement
+        if (variableTextAnchors.empty()) {
+            const auto placeFeature = [&](const CollisionFeature& collisionFeature,
+                                          style::TextWritingModeType orientation) {
+                textBoxes.clear();
```
Comment (tmpsantos) [inline src/mbgl/text/placement.cpp:317]: They do we clear placement cache again here and also one more time down this code?

File: src/mbgl/text/placement.cpp (hunk 2 above)
Comment (pozdnyakov) [reply src/mbgl/text/placement.cpp:146]: I need to keep them assignable for stable_sort (and I did not want to complicate this structure)

File: src/mbgl/text/placement.cpp (hunk 3 above)
Comment (pozdnyakov) [reply src/mbgl/text/placement.cpp:317]: yeah, one time is enough :smile:

**Comment by pozdnyakov** [comment]
@alexshalamov @tmpsantos I've extracted render tests changes to a separate pr https://github.com/mapbox/mapbox-gl-native/pull/16380 and considered your comments there, ptal.

**Comment by tmpsantos** [comment]
@tobrun I noticed that the Android tests are failing here, any reasons why?

File: src/mbgl/text/placement.cpp (hunk 3 above)
Comment (pozdnyakov) [reply src/mbgl/text/placement.cpp:317]: actually yes, and also at each variable placement attempt

File: src/mbgl/text/collision_index.cpp [hunk 4]
```diff
@@ -104,13 +104,42 @@ inline bool CollisionIndex::overlapsTile(const CollisionBoundaries& boundaries,
...
 }
 
-bool CollisionIndex::intersectsTileEdges(const CollisionBox& box,
-                                         Point<float> shift,
-                                         const mat4& posMatrix,
-                                         const float textPixelRatio,
-                                         const CollisionBoundaries& tileEdges) const {
-    auto collisionBoundaries = getProjectedCollisionBoundaries(posMatrix, shift, textPixelRatio, box);
-    return overlapsTile(collisionBoundaries, tileEdges) && !isInsideTile(collisionBoundaries, tileEdges);
+IntersectStatus CollisionIndex::intersectsTileEdges(const CollisionBox& box,
```
[the remark this answers is not in the document]
Comment (pozdnyakov) [reply src/mbgl/text/collision_index.cpp:107]: will add in a follow-up pr, planning to expose `IntersectStatus` in `PlacedSymbolData`

File: src/mbgl/text/placement.cpp [hunk 5]
```diff
@@ -1186,50 +1240,88 @@ void StaticPlacement::commit() {
...
+
+    std::stable_sort(intersections.begin(), intersections.end(), [](const Intersection& a, const Intersection& b) {
+        uint8_t flagsA = a.status.flags;
+        uint8_t flagsB = b.status.flags;
+        // Items arranged as: VerticalBorders & HorizontalBorders (3) ->
+        // VerticalBorders (2) -> HorizontalBorders (1)
+        if (flagsA != flagsB) return flagsA > flagsB;
+        // If both intersects the same border(s), look for a more noticable cut-off.
+        if (a.status.minSectionLength != b.status.minSectionLength) {
+            return a.status.minSectionLength > b.status.minSectionLength;
```
Comment (alexshalamov) [inline src/mbgl/text/placement.cpp:1304]: Should vertical and horizontal length be checked separately? Also, should it be total overlap area vs single 'section length'? Label that is in a tile corner can have bigger overlap area, yet, smaller section length.

Comment (pozdnyakov) [reply src/mbgl/text/placement.cpp:1304]: Thanks for the proposal! will consider it separately

**Pull request merged**